keps/sig-node/2008-forensic-container-checkpointing/README.md

```diff
@@ -292,14 +329,94 @@ Checkpointing containers will be possible again.
 
 ###### Are there any tests for feature enablement/disablement?
 
-Currently no.
+Currently the test will automatically be skipped if the feature is not enabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+Does not apply as the feature is an additional API endpoint with no
+dependencies on other functionality. If it is not enabled via the feature
+gate it will return `404 page not found`. If it is not enabled in the
+underlying container engine a `500` will be returned with an error
+message from the container engine. If it is enabled the API endpoint exists
+if disabled then it does not exist. No planning necessary.
+
+Documented at <https://kubernetes.io/docs/reference/node/kubelet-checkpoint-api/>
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+At this point it is still a kubelet only API endpoint and has no dependencies
+on other components.
+
+###### What specific metrics should inform a rollback?
+
+The only metric is the return code from the API endpoint.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No, this does not seem to apply for this feature.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+Querying the state of the feature gate offers the possibility to detect
+if the API endpoint will return `404` or not.
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+As it is not exposed in the Kubernetes API it cannot be determined. This is
+only visible in the kubelet.
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+###### How can someone using this feature know that it is working for their instance?
+
+The kubelet API endpoint can return following codes:
+
+- 200: checkpoint archive was successfully created
+- 404: feature is not enabled
+- 500: underlying container engine does not support checkpointing containers
+
+Documented at <https://kubernetes.io/docs/reference/node/kubelet-checkpoint-api/>
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+Does not apply as the enhancement will only be called when requested. Not a service.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+Does not apply as the enhancement will only be called when requested. Not a service.
```

Comment:
This section appears to apply since the capability surface are of the kubelet increases, a new endpoint has been added, new functionality behind that endpoint, and knowing its failure rate and performance characteristics are important.

Reply:
metrics can indicate the success % rate of the checkpointing API.

Reply:
AFAIK I don't believe we have any SLIs around the kubelet API. We are adding metrics to calculate the success/failure of the checkpoint CRI call.

But this is an optional feature and the main SLI/SLO I see related to this would be around pod lifecycle but this will not be impacted.

The SLI/SLO I looked at were here: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos

Reply:
I am open to adding metrics for the checkpointing API if that is possible. I added kubelet metrics above as something to be added to the code base.